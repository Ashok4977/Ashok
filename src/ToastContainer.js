import React, {createContext} from 'react'
import PropTypes from 'prop-types'
import Toast from './Toast'
import useToasts from './hooks/useToasts'

const ToastContext = createContext()

const ToastContainer = (props) => {
  const {addToast, getToastProps} = useToasts(props)
  const {toasts, ...toastProps} = getToastProps()

  return (
    <ToastContext.Provider value={addToast}>
      {props.children}
      {toasts &&
        toasts.map((toast) => {
          return <Toast key={toast.id} className={toast.className} toast={toast} {...toastProps} />
        })}
    </ToastContext.Provider>
  )
}

ToastContainer.Context = {
  addToast: PropTypes.func,
}

ToastContainer.propTypes = {
  autoDismiss: PropTypes.bool,
  timeout: PropTypes.number,
}
export default ToastContainer
