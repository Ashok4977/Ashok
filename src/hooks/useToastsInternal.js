import {nanoid} from 'nanoid'
import {useState} from 'react'

export const TOAST_ANIMATION_LENGTH = 300

const useToastsInternal = ({autoDismiss = true, timeout = 5000} = {}) => {
  const [toasts, setToasts] = useState([])

  const addToast = (freshToast) => {
    const toastId = nanoid()
    let timeoutId = 0
    if (autoDismiss) {
      timeoutId = window.setTimeout(startRemovingToast, timeout, toastId)
    }
    const newToast = {id: toastId, timeoutId, ...freshToast}
    // if there's already a toast on the page, wait for it to animate out before
    // adding a new toast
    if (toasts.length > 0) {
      const firstToast = toasts[0]
      startRemovingToast(firstToast.id)
      setTimeout(setToasts, TOAST_ANIMATION_LENGTH, [newToast])
      return
    }
    setToasts([newToast])
  }

  const cancelAutoDismiss = (toast) => {
    window.clearTimeout(toast.timeoutId)
  }

  // find the toast to remove and add the `toast-leave` class name
  // after the animation is run, the onAnimationEnd handler in Toast.js calls removeToast
  const startRemovingToast = (id) => {
    setToasts((prevState) => prevState.map((toast) => (toast.id === id ? {...toast, className: 'toast-leave'} : toast)))
    setTimeout(removeToast, TOAST_ANIMATION_LENGTH, id)
  }
  const findToastById = (id) => toasts.find((toast) => toast.id === id)
  const removeToast = (id) => {
    setToasts((currentToasts) =>
      currentToasts.filter((toast) => {
        if (autoDismiss && toast.id === id && toast.timeoutId) {
          window.clearTimeout(toast.timeoutId)
        }
        return toast.id !== id
      })
    )

    const currentToast = findToastById(id)
    if (currentToast?.onDismiss) {
      currentToast.onDismiss()
    }
  }

  const getToastProps = () => {
    return {toasts, removeToast, startRemovingToast, cancelAutoDismiss}
  }
  return {addToast, getToastProps}
}

export default useToastsInternal
