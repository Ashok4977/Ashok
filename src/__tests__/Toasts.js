import React from 'react'
import ToastContainer from '../ToastContainer'
import {render, cleanup, screen} from '@testing-library/react'
import {renderHook, act} from '@testing-library/react-hooks'
import useToastsInternal from '../hooks/useToastsInternal'

jest.useFakeTimers()

describe('Toast', () => {
  it.skip('Renders a toast on the page when it is added by the hook', () => {
    render(<ToastContainer />)
    const {result} = renderHook(() => useToastsInternal())
    act(() => {
      result.current.addToast({
        message: 'Changes Saved',
        type: 'success',
      })
    })

    expect(screen.getByRole('status')).toHaveTextContent('Changes Saved')
    cleanup()
  })

  it.skip('Prevents dismissing the toast when focused inside of it', () => {})

  it.skip('Adds an action item to the toast', () => {})

  it.skip('Adds only one action item to the toast', () => {})
})

describe('useToastsInternal', () => {
  it('Adds a toast to the container with addToast from the hook', () => {
    const {result} = renderHook(() => useToastsInternal())
    expect(result.current.getToastProps().toasts).toEqual([])

    act(() => {
      result.current.addToast({
        message: 'Changes Saved',
        type: 'success',
      })
    })

    expect(result.current.getToastProps().toasts.length).toEqual(1)
  })

  it('Adds only one toast at a time to the container', () => {
    const {result} = renderHook(() => useToastsInternal())
    expect(result.current.getToastProps().toasts).toEqual([])

    act(() => {
      result.current.addToast({
        message: 'Changes Saved',
        type: 'success',
      })
    })

    expect(result.current.getToastProps().toasts.length).toEqual(1)

    act(() => {
      result.current.addToast({
        message: 'Changes Saved',
        type: 'success',
      })
    })

    expect(result.current.getToastProps().toasts.length).toEqual(1)
  })

  it('Respects autodismiss: false', () => {
    const {result} = renderHook(() => useToastsInternal({autoDismiss: false}))

    act(() => {
      result.current.addToast({
        message: 'Changes Saved',
        type: 'success',
      })
    })

    expect(result.current.getToastProps().toasts[0].timeoutId).toEqual(undefined)
  })

  it('Adds timeout to new toasts by default', () => {
    const {result} = renderHook(() => useToastsInternal())

    act(() => {
      result.current.addToast({
        message: 'Changes Saved',
        type: 'success',
      })
    })

    expect(result.current.getToastProps().toasts[0].timeoutId).not.toBeNull()
  })
})
