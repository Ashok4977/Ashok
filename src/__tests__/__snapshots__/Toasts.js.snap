// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toasts renders consistently 1`] = `
<div
  className=""
/>
`;
